### ISSUE TITLE
[RFC] Rethink how plugins should be packaged and loaded

### ISSUE THREAD

**Issue by garyttierney** [apollo-rsps/apollo#337]
With the transition to Kotlin based plugins the need for maintaining lists of script files plugin.xml goes away.  Each `<x>.plugin.kts` file is compiled by the plugin compiler and added to a manifest (temporary) which is parsed by the server.

This new system solves a lot of the shortcomings of the old one.

1.  No need to declare plugin load order or dependencies (as long as they're on the classpath)
2.  Execution of plugin code is deferred until all plugins are loaded
3.  Plugins can safely declare public API code and share it among other plugins
4.  Support for stopping plugins when the server is going down

Currently the compiler splits plugin compilation into 2 stages.  1.  Compile plugin library and public API code (.kt files).  2.  Compile plugin scripts (.plugin.kts) files.  The output of these compile jobs is added to the servers classpath and a list of plugin script classes is generated for the server to load.  This should be split up so each plugin can be packaged individually (and include metadata on its authors and description).

**Comment by garyttierney** [comment]
A bit of thought on what a plugin tree might look like:

```
- test-plugin/
-- test-plugin/lib -- Directory containing public plugin code (no access to world or context in here)
---- test-plugin/lib/PublicClass.kt -- A public class which will be added to the classpath when this plugin is loaded
-- test-plugin/tests
---- test-plugin/tests/TestPluginIntegrationTest.kt -- Integration test to be executed for this plugin (test framework TODO)
-- test-plugin/test.plugin.kts -- Plugin entry point (can register event listeners and has access to the world + plugin context)
-- test-plugin/test.toml -- Plugin metadata (description, author, test suite information...)
```

**Comment by garyttierney** [comment]
This has been resolved as part of #338.  Plugins are simply Java class files that can be distributed using the same delivery mechanisms as any other Java artifacts.
